docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+# Support build and deployment of application container image using odo
+
+## Abstract
+Add a new command (verb) to build a production-like/slim container image for the project and deploy the built image on the target platform.
+
+There is an existing proposal for adding on outer-loop information (including support for multiple strategies for building and deploying the projects) to Devfile 2.1.0: 
+https://github.com/devfile/kubernetes-api/issues/49 
+
+It would be useful to start the design/development of a simpler version of `odo deploy` with devfile 2.0.0 that covers:
+- single build strategy - Dockerfile built using `buildah`
+- single deployment strategy - Kubernetes manifest deployed using `kubectl`.
+
+## Motivation
+`odo` is limited to development inner-loop for a project and there is no support for outer-loop - build a slim/production container image for your application and deploy the build container. It would be very useful for developers to be able to try inner-loop and then transition over to the outer-loop using odo. The outer-loop information could be provided by the application stack (devfile) to avoid developers having to worry about these aspects. 
+
+`odo deploy` can be a good way to assure the developer that their application can be built and deployed successfully using the build/deploy guidance that comes from devfile.
```

Comment:
You are right `odo push` also runs the app on k8s. During development, we use a base runtime image and sync the user's application code and run it. We do not create a container image for the application and deploy that application image. 

Typically for outer loop, you will build a container image for the application without any dev tools etc. and then deploy that built image. This can be significantly smaller and different to what we use in development. 

The proposal is allowing users to test that this flow will work based on info from devfile.

Reply:
Makes sense. Thank you. :+1: